group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/SubscribedTopicDescriberImpl.java

```diff
@@ -72,7 +71,8 @@ public int numPartitions(Uuid topicId) {
      */
     @Override
     public Set<String> racksForPartition(Uuid topicId, int partition) {
-        return Collections.emptySet();
+        TopicMetadata topic = this.topicMetadata.get(topicId);
+        return topic == null ? Set.of() : topic.partitionRacks().getOrDefault(partition, Set.of());
```

Comment:
I suppose that this won't work because the `topicMetadata` is not maintained in the group any more. Hence you not guaranteed to have it when the assignor is called. I was hoping that we could reuse the MetadataImage.